### PR TITLE
[WIP] Allow for external downloaders/converters

### LINKED ISSUE #213: fuel-download / fuel-convert for datasets not in the fuel distro
I have written a fairly comprehensive fuel adapter for the [Labeled Faces in the Wild dataset](http://vis-www.cs.umass.edu/lfw/). However, instead of asking to have this code merged into the fuel library, I'm asking that `fuel-download` and `fuel-convert` be extended to allow conversion of datasets that are not part of the fuel distribution itself.

I've put together a proof of concept for this. In my [fuel_lfw](https://github.com/dribnet/lfw_fuel) repo is [lfw.py](https://github.com/dribnet/lfw_fuel/blob/857dfb43bbcdeb5d820c95b541569af181617409/lfw_fuel/lfw.py) which includes a [`download_subparser` hook](https://github.com/dribnet/lfw_fuel/blob/857dfb43bbcdeb5d820c95b541569af181617409/lfw_fuel/lfw.py#L58) for fuel-download and a [`convert_subparser` hook](https://github.com/dribnet/lfw_fuel/blob/857dfb43bbcdeb5d820c95b541569af181617409/lfw_fuel/lfw.py#L172) for fuel-convert. Since `fuel-download` and `fuel-convert` are currently not currently intended for this use case I've also built my own short wrapper scripts called [`kero-download`](https://github.com/dribnet/kerosene/blob/23a480903519a3da1b2d379ebaba6004adf2566a/bin/kero-download) and [`kero-convert`](https://github.com/dribnet/kerosene/blob/23a480903519a3da1b2d379ebaba6004adf2566a/bin/kero-convert), which trick those scripts into working correctly. So using those, I can simply execute:

```bash
kero-download lfw_fuel/lfw.py
kero-convert lfw_fuel/lfw.py
```

to produce a fuel hdf5 file ready to be consumed by my blocks program.

In addition, the I provide an [`H5PYDataset` subclass](https://github.com/dribnet/lfw_fuel/blob/857dfb43bbcdeb5d820c95b541569af181617409/lfw_fuel/lfw.py#L182) so that other developers can easily use this dataset in their own blocks programs.

Separate from LFW, I have a handful of datasets that I would like to make available in a similar fashion.

So I'm hoping to get:
  * agreement from the fuel maintainers that this is a useful use-case to allow
  * functionality added to `fuel-download` and `fuel-convert` to support use-case (I can also help here)
  * some guidance on what the conventions should be for these 3rd party fuel datasets

Hopefully the fuel maintainers agree that this change makes fuel more powerful by allowing anyone to create and distribute datasets for the fuel ecosystem.

### PR DESCRIPTION
**Pull request by vdumoulin**
Fixes #213.

@dribnet the interface for adding external downloaders / converters would be to define an `extra_downloaders` / `extra_converters` config attribute in `~/.fuelrc` like so:

```
# ~/.fuelrc

extra_downloaders: ['my.extra.downloaders1', 'my.extra.downloaders2']
extra_converters: ['my.extra.converters1', 'my.extra.converters2']
```

where elements of `extra_downloaders` / `extra_converters` are modules which define an `all_downloaders` / `all_converters` attribute like in `fuel.downloaders` / `fuel.converters`.

Would that work for you?

### CODE REVIEW

**Comment by dribnet** [comment]
Wow - thanks for making forward progress on this so quickly.

Haven't reviewed the code yet - are these .fuelrc settings something that would have to be modified for each new dataset? I think an ideal developer flow would be something like:

```bash
pip install some_fuel_dataset
fuel-download some_fuel_dataset
fuel-convert some_fuel_dataset
```

Without any manual config edits.

**Comment by dribnet** [comment]
OK, just looked at the code. I think this is great - but to would suggest adding a command line argument to fuel-download and fuel-convert to allow items of fuel.config.extra_converters to also be specified on the command line as well. Then you at runtime you could just append this to the existing list in the .fuelrc. Thoughts?

**Comment by dribnet** [comment]
@vdumoulin - I've implemented this new command line argument as `--extra` for both `fuel-download` and `fuel-convert` and opened a merge request with this downstream commit branch at vdumoulin/fuel#1

I also made this [small tweak](https://github.com/dribnet/lfw_fuel/commit/82e7f756865d2e09b7b646b5b3045531b1b080f4) to my lfw_fuel project and confirmed that this implementation works as expected either by adding extra downloaders / converters to `~/.fuelrc`, or by using the new `--extra` argument. This is pushed up to my [download_convert branch in that repo](https://github.com/dribnet/lfw_fuel/commits/download_convert), and I'll paste in a sample session using that branch below as an example of testing various options:

```bash
# either `pip install` or just clone repo and update PYTHONPATH
export PYTHONPATH=`pwd`
```
```
$ fuel-download --extra lfw_fuel lfw
Downloading lfw.tgz, lfw-names.txt, pairsDevTest.txt, pairsDevTrain.txt

INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
lfw.tgz: 100% [======================================] Time: 0:00:27   6.62 MB/s
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
lfw-names.txt: 100% [================================] Time: 0:00:00 462.61 kB/s
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
pairsDevTest.txt: 100% [=============================] Time: 0:00:00  15.30 MB/s
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
pairsDevTrain.txt: 100% [============================] Time: 0:00:00 590.11 kB/s
```
```
$ fuel-convert --extra lfw_fuel lfw
--> Converting lfw.tgz to tar
--> Building test/train lists
--> Converting
Converting train images: 100% [==================================] Time: 0:00:13
Converting test images: 100% [===================================] Time: 0:00:05
train shapes:  (2200, 6, 250, 250) (2200, 1)
test shapes:   (1000, 6, 250, 250) (1000, 1)
--> Writing hdf5 output file
--> Done, removing tar file
```
```
$ fuel-convert --extra lfw_fuel lfw --format deepfunneled
usage: fuel-convert [-h] [--extra EXTRA]
                    
                    {iris,caltech101_silhouettes,svhn,cifar10,binarized_mnist,lfw,cifar100,mnist}
                    ...
fuel-convert: error: The following required files were not found:

   * lfw-deepfunneled.tgz

Did you forget to run fuel-download?
```
```
$ fuel-download --extra lfw_fuel --help
usage: fuel-download [-h] [--extra EXTRA]
                     
                     {iris,caltech101_silhouettes,svhn,cifar10,binarized_mnist,lfw,cifar100,mnist}
                     ...

Download script for built-in datasets.

positional arguments:
  {iris,caltech101_silhouettes,svhn,cifar10,binarized_mnist,lfw,cifar100,mnist}
    iris                Download the iris dataset
    caltech101_silhouettes
                        Download the caltech101_silhouettes dataset
    svhn                Download the svhn dataset
    cifar10             Download the cifar10 dataset
    binarized_mnist     Download the binarized_mnist dataset
    lfw                 Download the lfw dataset
    cifar100            Download the cifar100 dataset
    mnist               Download the mnist dataset

optional arguments:
  -h, --help            show this help message and exit
  --extra EXTRA         add an extra downloader package
```
```
$ fuel-download --extra lfw_fuel lfw --help
usage: fuel-download lfw [-h] [-d DIRECTORY] [--clear] [--format FORMAT]

optional arguments:
  -h, --help            show this help message and exit
  -d DIRECTORY, --directory DIRECTORY
                        where to save the downloaded files
  --clear               clear the downloaded files
  --format FORMAT       alternate format
```

**Comment by vdumoulin** [comment]
@dribnet I like your idea of being able to specify extra downloaders/converters from the command-line.

There are two ways to implement this:
* through an `--extra` command-line flag, like featured in your PR, or
* via environment variables that would override the `extra_downloaders` / `extra_converters` configuration variables, like `FUEL_DATA_PATH` does for `data_path`.

I think both options have their merits. On one hand, the `--extra` flag makes the feature more discoverable through `--help`. On the other hand, the environment variable is more consistent with how things are done in `fuel.config`.

Note that these to options aren't mutually exclusive either. I think we should strive for the most intuitive and discoverable interface.

What are your thoughts on this?

**Comment by factom** [comment]
Environment variables would also be great. Honestly I was a little unhappy with the incidental complexity of how I implemented `--extra` with the arg pre-processor, so would be happy to drop that version and switch instead to an environment variable based solution.

If I understand this proposal correctly, then my example goes from:
```bash
fuel-download --extra lfw_fuel lfw
```
to
```bash
EXTRA_DOWNLOADERS=lfw_fuel fuel-download lfw
```

Let me know if you'd like for me to rework my commit to do this.

**Comment by vdumoulin** [comment]
Your understanding is correct. In general, if you have more than one package you'd like to add, you can type something like

```
FUEL_EXTRA_DOWNLOADERS="package1 package2" fuel-download some_dataset
```

The fix is quite simple, so I went ahead and pushed something. Let me know if you spot anything wrong.

**Comment by dribnet** [comment]
@vdumoulin - confirmed to be working as of 73c986a, but extra_downloaders/converters then stopped working from both the .fuelrc and environment variable after the successive 0c5cbb0 commit. I'll try to dig into this more, but for now might want to hold off merging.

**Comment by dribnet** [comment]
Aha, like the cleaner subparser interface of returning the function. Updated my code and everything is working well again - confirmed operation both with environment variables and the .fuelrc file. 

So thanks for this great addition - now hoping this gets merged into master soon. Once merged, I'll update my downstream repo dependency and code, and retire my own hacked versions of `download` and `convert` which provided me this functionality.

Perhaps overkill, but below is an updated sample session (without .fuelrc entries) that demonstrates functionality.

```bash
# either `pip install` or just clone repo and update PYTHONPATH

export PYTHONPATH=`pwd`
```
```bash
# test extra downloader

$ FUEL_EXTRA_DOWNLOADERS="lfw_fuel" fuel-download lfw
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
lfw.tgz: 100% [======================================] Time: 0:00:26   6.71 MB/s
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
lfw-names.txt: 100% [================================] Time: 0:00:00 487.51 kB/s
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
pairsDevTest.txt: 100% [=============================] Time: 0:00:00  15.93 MB/s
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
INFO:requests.packages.urllib3.connectionpool:Starting new HTTP connection (1): vis-www.cs.umass.edu
pairsDevTrain.txt: 100% [============================] Time: 0:00:00  15.62 MB/s
```

```bash
# test extra converter

$ FUEL_EXTRA_CONVERTERS="lfw_fuel" fuel-convert lfw
--> Converting lfw.tgz to tar
--> Building test/train lists
--> Converting
Converting train images: 100% [==================================] Time: 0:00:13
Converting test images: 100% [===================================] Time: 0:00:05
train shapes:  (2200, 6, 250, 250) (2200, 1)
test shapes:   (1000, 6, 250, 250) (1000, 1)
--> Writing hdf5 output file
--> Done, removing tar file
```

```bash
# test both converter specific arguments and checks for missing files

$ FUEL_EXTRA_CONVERTERS="lfw_fuel" fuel-convert lfw --format deepfunneled
usage: fuel-convert [-h]
                    
                    {iris,caltech101_silhouettes,svhn,cifar10,binarized_mnist,lfw,cifar100,mnist}
                    ...
fuel-convert: error: The following required files were not found:

   * lfw-deepfunneled.tgz

Did you forget to run fuel-download?
```

```bash
# check top level help to confirm extra dataset gets listed

$ FUEL_EXTRA_DOWNLOADERS="lfw_fuel" fuel-download --help
usage: fuel-download [-h]
                     
                     {iris,caltech101_silhouettes,svhn,cifar10,binarized_mnist,lfw,cifar100,mnist}
                     ...

Download script for built-in datasets.

positional arguments:
  {iris,caltech101_silhouettes,svhn,cifar10,binarized_mnist,lfw,cifar100,mnist}
    iris                Download the iris dataset
    caltech101_silhouettes
                        Download the caltech101_silhouettes dataset
    svhn                Download the svhn dataset
    cifar10             Download the cifar10 dataset
    binarized_mnist     Download the binarized_mnist dataset
    lfw                 Download the lfw dataset
    cifar100            Download the cifar100 dataset
    mnist               Download the mnist dataset

optional arguments:
  -h, --help            show this help message and exit
```

```bash
# check that extra provided args are discoverable with help (note `format` argument)

$ FUEL_EXTRA_DOWNLOADERS="lfw_fuel" fuel-download lfw --help
usage: fuel-download lfw [-h] [-d DIRECTORY] [--clear] [--format FORMAT]

optional arguments:
  -h, --help            show this help message and exit
  -d DIRECTORY, --directory DIRECTORY
                        where to save the downloaded files
  --clear               clear the downloaded files
  --format FORMAT       alternate format
```

**Comment by vdumoulin** [comment]
@dribnet Sorry about that, I meant to write and push some documentation explaining the change but didn't have time to do so this weekend.

File: docs/built_in_datasets.rst [hunk 1]
```diff
@@ -13,8 +13,25 @@ to automate these operations.
...
+
+.. code-block:: yaml
+
+    # ~/.fuelrc
+    data_path: "/first/path/to/my/data:/second/path/to/my/data"
+
+When looking for data, Fuel will go through this sequence and use the first
+matching file it finds.
+
+This configuration variable can be overriden by setting the ``FUEL_DATA_PATH``
```
Comment (dwf) [inline docs/built_in_datasets.rst]: overridden

Comment (vdumoulin) [inline docs/built_in_datasets.rst]: 915e5d4

File: docs/built_in_datasets.rst [hunk 2]
```diff
@@ -124,3 +141,43 @@ argument:
...
+
+You should include the following in your ``~/.fuelrc``:
+
+.. code-block:: yaml
+
+    # ~/.fuelrc
+    extra_downloaders: ['package1.extra_downloaders', 'package2.extra_downloaders']
+    extra_converters: ['package1.extra_converters', 'package2.extra_converters']
+
+These configuration variables can be overriden through the
```
Comment (dwf) [inline docs/built_in_datasets.rst]: overridden

File: tests/test_config_parser.py [hunk 3]
```diff
@@ -1,9 +1,20 @@
...
+from fuel.config_parser import (Configuration, ConfigurationError,
+                                extra_downloader_converter)
+
+
+class TestExtraDownloaderConverter(object):
+    def test_iterable(self):
+        assert_equal(extra_downloader_converter(['a.b.c', 'd.e.f']),
+                     ['a.b.c', 'd.e.f'])
+
+    def test_str(self):
```
Comment (dwf) [inline tests/test_config_parser.py]: Test the case where we have just one module specified as string, as that is the sort of corner case that is going to break.

File: docs/built_in_datasets.rst [hunk 4]
```diff
@@ -13,8 +13,25 @@ to automate these operations.
...
+configuration variable has to be set inside ``~/.fuelrc``. It's expected to be
+a sequence of paths separated by an OS-specific delimiter (``:`` for Linux and
+OSX, ``;`` for Windows):
+
+.. code-block:: yaml
+
+    # ~/.fuelrc
+    data_path: "/first/path/to/my/data:/second/path/to/my/data"
+
+When looking for data, Fuel will go through this sequence and use the first
```
Comment (dwf) [inline docs/built_in_datasets.rst]:
```
When looking for a specific file (e.g. ``mnist.hdf5``, Fuel will search each of these paths in sequence, using the first matching file that it finds.
```

File: docs/built_in_datasets.rst [hunk 5]
```diff
@@ -124,3 +141,43 @@ argument:
...
+* ``package2.extra_downloaders``
+* ``package1.extra_converters``
+* ``package2.extra_converters``
+
+You should include the following in your ``~/.fuelrc``:
+
+.. code-block:: yaml
+
+    # ~/.fuelrc
+    extra_downloaders: ['package1.extra_downloaders', 'package2.extra_downloaders']
```
Comment (dwf) [inline docs/built_in_datasets.rst]: This works with the YAML list syntax too, e.g.
```
extra_downloaders:
- package1.extra_downloaders
- package2.extra_downloaders
extra_converters:
- package1.extra_converters
- package2.extra_converters
```
Might be more readable from a docs perspective.

File: docs/built_in_datasets.rst [hunk 6]
```diff
@@ -13,8 +13,25 @@ to automate these operations.
...
-environment variable has to be set.
+In order for Fuel to know where to look for its data, the ``data_path``
+configuration variable has to be set inside ``~/.fuelrc``. It's expected to be
+a sequence of paths separated by an OS-specific delimiter (``:`` for Linux and
+OSX, ``;`` for Windows):
+
+.. code-block:: yaml
+
+    # ~/.fuelrc
+    data_path: "/first/path/to/my/data:/second/path/to/my/data"
```
Comment (dwf) [inline docs/built_in_datasets.rst]: Just occurred to me that it would make a lot of sense to accept a list directly here too (a lot nicer to edit, with the syntax I showed you below), but we can leave that for another PR.

**Comment by dwf** [comment]
For my own edification (and for the purpose of preserving it for posterity in the PR comments), what's the motivation behind returning the downloader rather than setting it as `.func`?

LGTM aside from the issues I raised.

**Comment by vdumoulin** [comment]
I had some comments offline regarding the fact that passing the download/convert function via `set_defaults` is somewhat unconventional, and that it would be cleaner to have the function filling the subparser to return it instead.

File: docs/built_in_datasets.rst (hunk 6 above)
Comment (vdumoulin) [inline docs/built_in_datasets.rst]: I agree with you. Since this involves changing documentation in Blocks as well, I created a ticket for that (#218), and I'll address it later.

File: docs/built_in_datasets.rst (hunk 2 above)
Comment (vdumoulin) [inline docs/built_in_datasets.rst]: b0a224f

File: tests/test_config_parser.py (hunk 3 above)
Comment (vdumoulin) [inline tests/test_config_parser.py]: 28d9cf8

**Comment by vdumoulin** [comment]
@dwf I addressed your comments.
